### ISSUE TITLE
Why does not stop gradient for a?

### ISSUE THREAD

**Issue by khanrc** [bioinf-jku/coulomb_gan#1]
Hi, thanks for this interesting work. I recently implemented CoulombGAN in my project (https://github.com/khanrc/tf.gans-comparison) and I had two questions.

When calculating the potential, stop the gradients only for x and y. I think we should stop gradients for a too, isn't it?

Also, is there no convergence measure? I expected that the potential of fake samples serve as a convergence measure, but it did not.

**Comment by untom** [comment]
`a` is the parameter through which gradients flow back to the GAN. We stop the gradient for x and y, because otherwise when we calculate  `potential(x, x)`, we would backpropagate through both parameters, which we do not want. Think of it this way: each point in `x` and `y` is used twice in the potential function: once as "charge" used to create the potential field, and once as location of evaluation (this is `a`). This later point is actually the one we want to backpropagate through.

As for convergence measure: I use FID as convergence measure and monitor it to decide when to stop training (this is done manually). The potential of the fake samples could also been used, but since FID is the end goal I want to optimize for anyways, it makes sense to monitor that instead.

**Comment by khanrc** [comment]
Thanks for your quick and kind reply!

I thought back to the answer, is `stop_gradient` required when calculating the potential in the first place? The variables `x` and `y` which are used to compute the potential are independent of the discriminator and the variables are related only to the generator, but the potential is used only in the discriminator loss. So I thought that the gradient in the potential calculation will not delivered to the variables of G and D anyway. 

If I thought I was wrong somewhere, please let me know where I am wrong.
Thanks again for the answer.

**Comment by untom** [comment]
You are right, they aren't needed the way everything is set up, but e.g. if one would forget to filter out which variables to update when learning the discriminator (see [here](https://github.com/bioinf-jku/coulomb_gan/blob/master/train.py#L101)) one could end up updating the generator parameters using gradients from the discriminator loss by mistake. So I thought better be safe than sorry ;)

**Comment by khanrc** [comment]
I take your point. It was a pleasant conversation :)
Thank you!

**Comment by untom** [comment]
You're welcome, and best of luck with your project! :)
